src/main/java/org/jenkinsci/plugins/workflow/cps/FlowHead.java

```diff
@@ -114,29 +116,32 @@ void newStartNode(FlowStartNode n) throws IOException {
         execution.storage.storeNode(head, false);
     }
 
-    void setNewHead(FlowNode v) {
+    void setNewHead(@Nonnull FlowNode v) {
+        if (v == null) {
+            throw new IllegalArgumentException("FlowHead.setNewHead called on FlowHead id="+this.id+" with a null FlowNode, execution="+this.execution);
+        }
         try {
             if (this.head != null) {
                 CpsFlowExecution.maybeAutoPersistNode(head);
             }
             execution.storage.storeNode(v, true);
-            v.addAction(new TimingAction());
-            this.head = v;
-            CpsThreadGroup c = CpsThreadGroup.current();
-            if (c !=null) {
-                // if the manipulation is from within the program executing thread, then
-                // defer the notification till we get to a safe point.
-                c.notifyNewHead(v);
-            } else {
-                // in recovering from error and such situation, we sometimes need to grow the graph
-                // without running the program.
-                // TODO can CpsThreadGroup.notifyNewHead be used instead to notify both kinds of listeners?
-                execution.notifyListeners(Collections.singletonList(v), true);
-                execution.notifyListeners(Collections.singletonList(v), false);
-            }
         } catch (IOException e) {
             LOGGER.log(Level.FINE, "Failed to record new head: " + v, e);
         }
+        v.addAction(new TimingAction());
```

Comment:
So the idea here is that we want to still do the `TimingAction`, `notifyNewHead`, etc, even if we can't store the node?